### ISSUE TITLE
Double check calculation of PIC hashes for functions

### ISSUE THREAD

**Issue by danielplohmann** [danielplohmann/smda#37]
Looking at the example provided in the IDA blog post for [Diaphora](https://hex-rays.com/blog/plugin-focus-diaphora/), the two DLLs are almost identical, which should lead to many overlapping PIC hashes.
Given binaries:
* 408cb1604d003f38715833a48485b6a4e620edf163fb59aef792595866e4796b
* c115d15807b96dcb9871ebc69618ef77473f1451c427e7349f9aa3c72891ddc2
this is however not the case, e.g. function 0x633b3b81 (both binaries, same offset) appears identical in both binaries, yet produces two distinct PIC hashes?
* 0x304cda7a0ed10b98
* 0xaa2ceb7c973bc0c5

**Comment by fishfacegit** [comment]
The 'picblockhash' of escaped blocks should always be identical of two binaries since a block is executed in order.
However calculate a pic hash on a function the blocks must be in the correct order, so the problem could be the sorting of blocks in your code. The actual control flow is broken in basic blocks by escaping them. So simple block swapping could break your PIC hash on functions.

```
def _calculatePicHash(self, binary_info):
        escaped_binary_seqs = []
        for _, block in sorted(self.blocks.items()):
            for instruction in block:
                escaped_binary_seqs.append(instruction.getEscapedBinary(self._escaper, lower_addr=binary_info.base_addr, upper_addr=binary_info.base_addr + binary_info.binary_size))
        as_bytes = bytes([ord(c) for c in "".join(escaped_binary_seqs)])
        return struct.unpack("Q", hashlib.sha256(as_bytes).digest()[:8])[0]
```

To resolve this you could:

1. try to recreate the control flow by reordering basic blocks for a specific function
2. ordering resistant approach: like creating a histogram of blocks where the order does not matter

Could you provide the binaries that issue the mismatch?

**Comment by danielplohmann** [comment]
Hi!

Thanks for your comment!
RIght now, the blocks are already sorted based on their address, as happens in line 
```python
for _, block in sorted(self.blocks.items()):
```
in your code snippet.
Your comment made me revisit the situation and I have identified two issues so far.

First, I SMDA has not been wildcarding intraprocedural jumps, but only interprocedural ones, because I was blindly assuming that PIC-identical functions will most likely have the same layout and spacing between blocks.
I have already found functions now where this assumption does not hold true and blocks have been dislocated versus each other in similar yet not identical binaries.
This is easily fixed by simply passing the respective argument in `SmdaFunction`.

Second, SMDA currently only masks out bytes for a subset of address-referring data manipulating (`mov` etc) instructions, namely such, where just a pointer is directly used and not additionally displaced by calculation using registers etc.
However, this also messes with PIC matches and should be appropriately addressed.
This occurs e.g. here 
```
in 408cb1604d003f38715833a48485b6a4e620edf163fb59aef792595866e4796b
633B8F5D | 0F B6 81 80 8F 3B 63                          movzx   eax, ds:byte_633B8F80[ecx]
---
vs
---
in c115d15807b96dcb9871ebc69618ef77473f1451c427e7349f9aa3c72891ddc2
633B926D | 0F B6 81 90 92 3B 63                          movzx   eax, ds:byte_633B9290[ecx]
```

I think I would want to look into this to come up with a final solution that needs no further improvements, as it means recalculation of indices for [MCRIT](https://github.com/danielplohmann/mcrit).

[binaries.zip](https://github.com/danielplohmann/smda/files/12858265/binaries.zip)
I've added the example binaries for further study.

**Comment by malwarefrank** [comment]
You may be able to learn from what CMU/SEI has implemented. They have been using Position Independent Code (PIC) hashing for a number of years, first with their [Pharos framework](https://github.com/cmu-sei/pharos/blob/master/tools/fn2hash/fn2hash.pod) and then with their [Kaiju plugin](https://github.com/CERTCC/kaiju/blob/main/src/main/java/kaiju/tools/fnhashclassic/UnitHashData.java) for Ghidra.  They calculate both a PIC hash and what they call a Composite PIC hash.  The Composite PIC hash appears to be the same as their PIC one except that it ignores all control flow instructions.

If I am reading their code correctly, it looks like they overwrite address operands with NULLs/zeros if it references an address outside the current "chunk".  But I am uncertain whether "chunk" refers to the basic block, function, or some other boundary.

**Comment by danielplohmann** [comment]
@malwarefrank thanks a lot for pointing to the concrete locations in the original source code! 
My implementation so far was really just based on the original (written) publication and how I thought it was meant to be used. I just decided to use the hex representation and `??` wildcards as input for hashing directly, because that was already what could be used by YARA if I ever would have decided to expose that in `SmdaFunction`.

Ultimately, I think I already got somewhat close to the original as it works pretty fine in its current state but seems like I won't get around finally building some regressions to really cover everything appropriately and achieve what is supposed to do in SMDA.
The way it's done in Pharos and Ghidra is that they rely on detailed output from the disassembler engines to locate the offsets in the byte representation, which I frowned upon because I was always eyeing possibly supporting Zydis as a drop-in replacement for Capstone. However, that certainly didn't simplify the wildcarding. :smile: 

Also very entertaining to me, e.g. in the [Pharos part](https://github.com/cmu-sei/pharos/blob/f29bada3f84a67c393ae2acd86b0b0358c7767dd/libpharos/funcs.cpp#L1055) so many of the verbose comments in their source code and the implied contemplations hit home a lot as I relived these trains of thought. Will definitely study that code again when updating SMDA.

**Comment by danielplohmann** [comment]
Alright, I have revisited the implementation from Pharos and mkYARA. 

For that I had to dive deeper into capstone and understand its possibilities to derive operand offsets within instructions and figured out that it could not provide what I was looking for (or rather produced unexpected output, which is why I'm sticking with the "search" approach instead of using intrinsics provided by the disassembler - sadly no AST style examination of instruction semantics like in Pharos/Rose).

However, that way I also found out that mkYARA actually has a bug in its wildcarding. In case the opcode byte starts with "00", then, mkYARA will also wildcard this byte in addition to operand bytes (which it should not as it preserves opcode bytes for all other possibilities). The impact is however negligible as the corresponding instruction is a rather rare form of "add, X, X". Based on statistics I made some years ago, this instructions should occur in less than 1% of cases.

Ultimately, the issues above should now be wildcarded properly, as multi-byte operand jumps within functions are addressed by using the intraprocedural wildcarding feature and I also added some more wildcarding for possible occurrences of immediates.
Good enough for me for now, otherwise any future issues will be then covered with additional unit tests.
